diffusion/datasets/image_caption.py

```diff
@@ -0,0 +1,208 @@
+
+import torch
+from PIL import Image
+from streaming import Stream, StreamingDataset
+from torch.utils.data import DataLoader
+from torchvision import transforms
+from transformers import AutoTokenizer
+
+from diffusion.datasets.laion.transforms import LargestCenterSquare
+
+# Disable PIL max image size limit
+Image.MAX_IMAGE_PIXELS = None
+
+
+class StreamingImageCaptionDataset(StreamingDataset):
+    """Streaming dataset for image-caption pairs.
+
+    Args:
+        streams (Sequence[Stream], optional): One or more Streams to stream/cache samples from.
+            ``StreamingImageCaptionDataset`` uses either ``streams`` or ``remote``/``local``. Default:``None``.
+        remote (str, optional): Remote directory (S3 or local filesystem) where dataset is stored. Default: ``None``.
+        local (str, optional): Local filesystem directory where dataset is cached during operation. Default: ``None``.
+        split (str, optional): The dataset split to use. Currently, only ``None`` is supported. Default: ``None``.
+        shuffle (bool): Whether to shuffle the samples in this dataset. Default: ``False``.
+        tokenizer_name_or_path (str): The name or path of the tokenizer to use. Default: ``'stabilityai/stable-diffusion-2-base'``.
+        caption_selection (str): If there are multiple captions, specifies how to select a single caption.
+            'first' selects the first caption in the list and 'random' selects a random caption in the list.
+            If there is only one caption, this argument is ignored. Default: ``'first'``.
+        transform (Optional[Callable]): The transforms to apply to the image. Default: ``None``.
+        image_size (Optional[int]): The size to resize the image to. Default: ``None``.
+        image_key (str): Key associated with the image in the streaming dataset. Default: ``'image'``.
+        caption_key (str): Key associated with the caption in the streaming dataset. Default: ``'caption'``.
+        **streaming_kwargs: Additional arguments to pass in the construction of the StreamingDataloader
+    """
+
+    def __init__(
+        self,
+        streams: Optional[Sequence[Stream]] = None,
+        remote: Optional[str] = None,
+        local: Optional[str] = None,
+        split: Optional[str] = None,
+        shuffle: bool = False,
+        tokenizer_name_or_path: str = 'stabilityai/stable-diffusion-2-base',
+        caption_drop_prob: float = 0.0,
+        caption_selection: str = 'first',
+        transform: Optional[Callable] = None,
+        image_size: Optional[int] = None,
+        image_key: str = 'image',
+        caption_key: str = 'caption',
+        **streaming_kwargs,
+    ) -> None:
+
+        super().__init__(
+            streams=streams,
+            remote=remote,
+            local=local,
+            split=split,
+            shuffle=shuffle,
+            **streaming_kwargs,
+        )
+
+        if caption_selection not in ['first', 'random']:
```

Comment:
Probably should do the `.lower()` here before the check

Reply:
🤦